doc/specs/#976 - VT52 escape sequences.md

```diff
@@ -0,0 +1,141 @@
+---
+author: James Holderness @j4james
+created on: 2019-07-17
+last updated: 2019-07-17
+issue id: 976
+---
+
+# VT52 Escape Sequences
+
+## Abstract
+
+This spec outlines the work required to split off the existing VT52 commands from the VT100 implementation, and extend the VT52 support to cover all of the core commands.
+
+## Inspiration
+
+The existing VT52 commands aren't currently implemented as a separate mode, so they conflict with sequences defined in the VT100 specification. This is blocking us from adding support for the VT100 Index (IND) escape sequence, which is one of the missing commands required to pass the test of cursor movements in Vttest.
+
+## Solution Design
+
+The basic idea is to add support for the [DECANM private mode sequence](https://vt100.net/docs/vt100-ug/chapter3.html#DECANM), which can then be used to switch from the default _ANSI_ mode, to a new _VT52_ mode.
+
+In terms of implementation, there are a number of areas of the system that would need to be updated.
+
+### The State Machine
+
+In order to implement the VT52 compatibility mode correctly, we'll need to introduce a flag in the `StateMachine` class that indicates the mode that is currently active. When in VT52 mode, certain paths in the state diagram should not be followed - for example, you can't have CSI, OSC, or SS3 escape sequences. There would also need to be an additional state to handle VT52 parameters (for the _Direct Cursor Address_ command). These parameters take a different form to the typical VT100 parameters, as they follow the command character instead of preceding it.
+
+It would probably be best to introduce a new dispatch method in the `IStateMachineEngine` interface to handle the parsed VT52 sequences, since the existing `ActionEscDispatch` does not support parameters (which are required for the _Direct Cursor Address_ command). I think it would also make for a cleaner implementation to have the VT52 commands separate from the VT100 code, and would likely have less impact on the performance that way.

```

Comment:
I did initially try implementing everything in the existing `ActionEscDispatch`, but it felt messy to me, and it means we'd have to track the mode in the `OutputStateMachineEngine` as well, which becomes a bit of a pain to initialise.

And the state machine knows exactly want it wants the output engine to do, so it feels a bit convoluted to not call the code directly. Instead we'd now have to set a flag first, and then call a generic method that checks that flag to finally figure out what must actually be done.

I don't feel that strongly about it - and it's possible I'm not fully aware of the downsides - but I do think a separate `ActionVt52EscDispatch` would be preferable.

Reply:
Hm. Perhaps we preserve `ActionEscDispatch` as a router/decision mechanism between `ActionVt52EscDispatch` and `ActionVt100EscDispatch`, move the current impl into the 100 one, and just decide at this entrypoint.

I'd want @zadjii-msft's opinion here as he's done a lot more restructuring of the VT classes in the recent era than I have.

Reply:
In my opinion, I think I like it the way you have it in the spec currently. Not having to keep an extra piece of redundant state in the OutputStateMachineEngine does seem cleaner to me. 

I guess my main question is how does the OutputStateMachineEngine communicate to the StateMachine that it wants parameters too? Wouldn't the state machine have to know specifically that `ESC Y` in VT52 mode is followed by parameters, while everything else isn't?

Reply:
That's a good question. It feels bit hacky, but that knowledge would have to be part of the `StateMachine` itself. Similar to the way you have methods like `s_IsOscTerminator`, I was going with a `s_IsVt52ParamCommand` method (only true for `Y`) which would determine whether the parser needed to move to the new VT52 parameter state instead of dispatching. It's unfortunate that there isn't a neat ASCII range that you could say was reserved for parameter commands, like you have intermediates and finals in VT100+, but that's just the way it is.